homeassistant/components/buienradar/camera.py

```diff
@@ -0,0 +1,176 @@
+            self.async_camera_image(), self.hass.loop).result()
+
+    def __needs_refresh(self) -> bool:
+        if not (self._delta and self._deadline and self._last_image):
+            return True
+
+        return dt_util.utcnow() > self._deadline
+
+    async def __retrieve_radar_image(self) -> bool:
+        """Retrieve new radar image and return whether this succeeded."""
+        session = async_get_clientsession(self._hass)
+
+        url = RADAR_MAP_URL_TEMPLATE.format(w=self._dimension,
+                                            h=self._dimension)
+
+        if self._last_modified:
+            headers = {'If-Modified-Since': self._last_modified}
+        else:
+            headers = {}
+
+        try:
+            async with session.get(url, timeout=5, headers=headers) as res:
+                if res.status == 304:
+                    _LOG.debug("HTTP 304 - success")
+                    return True
+                if res.status != 200:
+                    _LOG.error("HTTP %s - failure", res.status)
+                    return False
+
+                last_modified = res.headers.get('Last-Modified', None)
+                if last_modified:
+                    self._last_modified = last_modified
+
+                self._last_image = await res.read()
+                _LOG.debug("HTTP 200 - Last-Modified: %s", last_modified)
+
+                return True
+        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
+            _LOG.error("Failed to fetch image, %s", type(err))
+            return False
+
+    async def async_camera_image(self) -> Optional[bytes]:
+        """
+        Return a still image response from the camera.
+
+        Uses ayncio conditions to make sure only one task enters the critical
+        section at the same time. Othertwise, two http requests would start
+        when two tabs with home assistant are open.
+
+        An additional boolean (_loading) is used to indicate the loading status
+        instead of _last_image since that is initialized to None.
+        """
+        if not self.__needs_refresh():
+            return self._last_image
+
+        # get lock, check iff loading, await notification if loading
+        async with self._condition:
+            if self._loading:
+                _LOG.debug("already loading - waiting for notification")
+                await self._condition.wait()
+                return self._last_image
+
+            # Set loading status **while holding lock**, makes other tasks wait
+            self._loading = True
+
+        try:
+            now = dt_util.utcnow()
+            res = await self.__retrieve_radar_image()
+            # successful response? update deadline to time before loading
+            if res:
+                self._deadline = now + timedelta(seconds=self._delta)
+
+            return self._last_image
+        finally:
+            # get lock, unset loading status, notify all waiting tasks
+            async with self._condition:
```

Comment:
If one request is waiting for the result, it does so inside the `async with self._condition`. Doesn't that mean it never releases the lock ? And if so, we will never be able to acquire the lock here?

Reply:
There could be deadlocks in here if I missed something (especially with exceptions) but this is not one 🙂

`await self._condition.wait()` releases the lock and acquires it before continuing. I updated the docstring to clarify this

Reply:
Oh, TIL.